### ISSUE TITLE
[question] Best practice to use with a RabbitMQ HA cluster

### ISSUE THREAD

**Issue by ebuildy** [pdezwart/php-amqp#215]
Sorry to ask question here, just wondering what would be the best practice to use this Php extension with a 3 nodes RabbitMQ HA cluster with 1 mirrored queue.

Currently I am using HAProxy in front of these, so Php connects to HAProxy then HAProxy load balance to different nodes. This is working very fine, but a member of my team asks if we could remove HAProxy.

1- Can we specify all nodes as host? ``new AMQPConnection(['host' => '172.20.0.2,172.20.0.3,172.20.0.4']);``)

2- Do you think HAProxy (hence a single IP) could bring problem?

Thanks you very much for your time,

Tom

**Comment by pinepain** [comment]
Hi. I also used HAProxy in the same manner as you and it was fine. Later there was ucarp configured to add failover for HAProxy machine, but I even not sure that it was brought to production. For some reason, mostly organizational we didn't used round-robin DNS, but it should be also fine to balance load with it, sure, it less flexible manner.

1. There are no such feature to specify multiple separate hosts within single `AMQPConnection`, but you can have multiple `AMQPConnection`s or better pick random host and connect to it on start and on failure fallback to another. In this case you don't even need a HAProxy. You can even implement server weight here. 

  There are [Connecting to Clusters from Clients](https://www.rabbitmq.com/clustering.html#clients) manual section in [Clustering Guide](https://www.rabbitmq.com/clustering.html) which describe pros and cons of this solution.

2. If it can fail, it will fail. But maybe not, it depends. If it fail and your apps can survive and not data lost, then it might be fine, otherwise it's all up to you. I had both single AMQP entry point and duplicated and had various kind of failures, including wildcard firewall rule was added to router so all RabbitMQ hosts, including three HAProxy machines was inaccessible, or HAProxy cluster was shutdown by mistake, so it's very broad question whether single HAProxy can cause trouble. If you are capable to add some redundancy, it may save your ass one time or even save money for business.

  Maybe, as suggested in clustering guide, you may add round-robin DNS record for RabbitMQ machines so no HAProxy then required, but here DNS server become a single point of failure instead or HAProxy (in case you don't have multiple DNS). Using DNS round-robin for failover may require adding some reconnect attempt, but it normally shouldn't be necessary, while we use in php-amqp [rabbitmq-c](https://github.com/alanxz/rabbitmq-c) C library, which under the hood try to connect to all host addresses one by one:

  from [rabbitmq-c's amqp_socket.c:449](https://github.com/alanxz/rabbitmq-c/blob/master/librabbitmq/amqp_socket.c#L449):

   ```C
  struct addrinfo *address_list;
  struct addrinfo *addr;

// ...

  last_error = getaddrinfo(hostname, portnumber_string, &hint, &address_list);

  if (0 != last_error) {
    return AMQP_STATUS_HOSTNAME_RESOLUTION_FAILED;
  }

  for (addr = address_list; addr; addr = addr->ai_next) {
    if (-1 != sockfd) {
      amqp_os_socket_close(sockfd);
    }

    sockfd = amqp_os_socket_socket(addr->ai_family, addr->ai_socktype, addr->ai_protocol);

    if (-1 == sockfd) {
      last_error = AMQP_STATUS_SOCKET_ERROR;
      continue;
    }
    // ... and so on
   ```

You may also ask this question in [rabbitmq-users mailing list](https://groups.google.com/forum/#!forum/rabbitmq-users), there are support engineers and RabbitMQ engineers subscribed to that mailing list and it is pretty active, maybe you question is already covered by older answers, give it a look.

Hope, it will help.

**Comment by KIVagant** [comment]
> and on failure fallback to another

This behaviour can bring to timeout failures, I got a similar problem with Nginx upstreams. One dead host can kill important production feature.

**Comment by pinepain** [comment]
@KIVagant yeah, [Consul DNS](https://www.consul.io/docs/agent/dns.html) fits here perfectly. 

There are even [rabbitmq-autocluster](https://github.com/aweber/rabbitmq-autocluster):
> A RabbitMQ plugin that clusters nodes automatically using Consul, etcd2, or DNS  

which I didn't tried yet, but it looks very promising.

**Comment by ebuildy** [comment]
Woa, many thanks for a such good answer as usually here, really appreciate your help.

Didn't think about DNS round robin, it seems a very good idea indeed. Will give a try this morning.

I send message to RabbitMQ after called ``fastcgi_finish_request`` so performance is not a problem here.

**Comment by c-datculescu** [comment]
Hello. Just chipping in. 

What we currently use in production is either a hardware load balancer or a software one (haproxy) in front of all the RabbitMQ nodes.

Regarding the fcgi_finish_request part, we did run into a issue where we were blocking the fpm threads after the output has been sent to client, leading to the entire pool being consumed. Our solution for this was to actually have a decently small connect_timeout (in case something fails).

We prefer to use load balancers because of their ability to do health checking, which means that they can remove the nodes as soon as they are detected as being down (by whatever check you use). Also, we prefer them because on most of them you can retry after a failed attempt on another node, which gives quite a lot of flexibility.

Also for load balancers becoming a single point of failure, there are solutions like the one mentioned above (probably involving virtual ips. i think ucarp uses the same thing). We never had a major issue in one year and something of operation with this setup.

**Comment by ebuildy** [comment]
Interesting, i didnt pay  attention to the  php  workers pool, sometime i
see  lot of connections opens between PHP and rabbitmq, maybe a bug with
persistance setting ?

Hello. Just chipping in.

What we currently use in production is either a hardware load balancer or a
software one (haproxy) in front of all the RabbitMQ nodes.

Regarding the fcgi_finish_request part, we did run into a issue where we
were blocking the fpm threads after the output has been sent to client,
leading to the entire pool being consumed. Our solution for this was to
actually have a decently small connect_timeout (in case something fails).

We prefer to use load balancers because of their ability to do health
checking, which means that they can remove the nodes as soon as they are
detected as being down (by whatever check you use). Also, we prefer them
because on most of them you can retry after a failed attempt on another
node, which gives quite a lot of flexibility.

Also for load balancers becoming a single point of failure, there are
solutions like the one mentioned above (probably involving virtual ips. i
think ucarp uses the same thing). We never had a major issue in one year
and something of operation with this setup.

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub
<https://github.com/pdezwart/php-amqp/issues/215#issuecomment-210482729>

**Comment by c-datculescu** [comment]
Generally we try to avoid using the persistent connections since php is known to not handle them very well. We use regular connections for both publishing and consuming from the nodes.

We did run into half tcp sockets opened, especially after we had a crash on rabbitmq (happened some months ago), where it seemed mnesia didn't clear the connections internally, but the ports were still available on the rabbitmq machine. We had to manually close down the connections and it all got fixed.
